### PR TITLE
Adding random forest

### PR DESCRIPTION
**Pull request by romanie**

### CODE REVIEW

File: diplom.py [hunk 1]
```diff
@@ -1,4 +1,12 @@
 #!/usr/bin/python
+import ml_metrics as metrics
+import argparse
+
+parser = argparse.ArgumentParser()
+parser.add_argument("--model", choices=["random_forest", "logistic_regression"], default=["random_forest"], dest="model")
+args = parser.parse_args()
+print args.model
+
 print "hello"
```
Comment (romanie) [inline diplom.py]: Remove this debug output

File: diplom.py [hunk 2]
```diff
@@ -1,4 +1,12 @@
 #!/usr/bin/python
+import ml_metrics as metrics
+import argparse
+
+parser = argparse.ArgumentParser()
+parser.add_argument("--model", choices=["random_forest", "logistic_regression"], default=["random_forest"], dest="model")
+args = parser.parse_args()
+print args.model
```
Comment (romanie) [inline diplom.py]: Please write a richer logging message, e.g.: 
print "Model used in this run is %s" % args.model

File: diplom.py [hunk 3]
```diff
@@ -53,32 +61,51 @@ def mean_average_precision(predicted_probability_distribution,
 # print test_X[0], test_y[0]
 
 from sklearn import linear_model
-logistic = linear_model.LogisticRegression(C = 1, solver = 'lbfgs', 
-    multi_class = 'multinomial', max_iter = 100)
-print('LogisticRegression score: %f'
-      % logistic.fit(train_X, train_y).score(test_X, test_y))
+from sklearn.ensemble import RandomForestClassifier
```
Comment (romanie) [inline diplom.py]: Could you move all the import to the header of the file?

File: diplom.py [hunk 4]
```diff
@@ -53,32 +61,51 @@ def mean_average_precision(predicted_probability_distribution,
...
+if args.model == "logistic_regression":
+    logistic = linear_model.LogisticRegression(C = 1, solver = 'lbfgs', 
+                                               multi_class = 'multinomial', max_iter = 100)
+else:
+    logistic = RandomForestClassifier(n_estimators=10)
+    
+logistic.fit(train_X, train_y)
+# print('LogisticRegression score: %f'
```
Comment (romanie) [inline diplom.py]: Remove this comment.

File: diplom.py [hunk 5]
```diff
@@ -53,32 +61,51 @@ def mean_average_precision(predicted_probability_distribution,
...
-logistic = linear_model.LogisticRegression(C = 1, solver = 'lbfgs', 
-    multi_class = 'multinomial', max_iter = 100)
-print('LogisticRegression score: %f'
-      % logistic.fit(train_X, train_y).score(test_X, test_y))
+from sklearn.ensemble import RandomForestClassifier
+if args.model == "logistic_regression":
+    logistic = linear_model.LogisticRegression(C = 1, solver = 'lbfgs', 
+                                               multi_class = 'multinomial', max_iter = 100)
+else:
+    logistic = RandomForestClassifier(n_estimators=10)
```
Comment (romanie) [inline diplom.py]: I think you'd need n_estimators=100

File: diplom.py [hunk 6]
```diff
@@ -53,32 +61,51 @@ def mean_average_precision(predicted_probability_distribution,
...
+    
+logistic.fit(train_X, train_y)
+# print('LogisticRegression score: %f'
+#       % logistic.fit(train_X, train_y).score(test_X, test_y))
 
 predicted_probability_distribution_test = logistic.predict_proba(test_X)
 predicted_probability_distribution_train = logistic.predict_proba(train_X)
 
+classes_by_probability = []
```
Comment (romanie) [inline diplom.py]: Please make decision whether you are using you own function or the library one.

File: diplom.py [hunk 7]
```diff
@@ -53,32 +61,51 @@ def mean_average_precision(predicted_probability_distribution,
...
+                                   logistic.classes_, 
+                                   test_y)
 print MAP_test
 MAP_train = mean_average_precision(predicted_probability_distribution_train, 
-                                   logistic.classes_, 
-                                   train_y) 
+                                    logistic.classes_, 
+                                    train_y) 
 print MAP_train
 # from sklearn import svm
```
Comment (romanie) [inline diplom.py]: Lets cleanup this part if you don't need it anymore.

File: diplom.py [hunk 8]
```diff
@@ -53,32 +61,51 @@ def mean_average_precision(predicted_probability_distribution,
...
+    for i in xrange(0, len(logistic.classes_)):
+        classes_by_prob[i] = classes_by_prob[i][1]
+    classes_by_probability.append(classes_by_prob)
+
+print metrics.mapk([[y] for y in test_y], classes_by_probability, k=5)
 
 MAP_test = mean_average_precision(predicted_probability_distribution_test, 
-                                  logistic.classes_, 
-                                  test_y)
+                                   logistic.classes_,
```
Comment (romanie) [inline diplom.py]: abnormal indent, revert this change please (probably you have used some tabs again).

File: diplom.py [hunk 9]
```diff
@@ -19,7 +27,7 @@ def mean_average_precision(predicted_probability_distribution,
         classes_by_prob = []
         for i in xrange(0, len(list_of_classes)):
             classes_by_prob.append([row[i], list_of_classes[i]])
-            classes_by_prob.sort(reverse=True)    
+        classes_by_prob.sort(reverse=True)    
         for j in range (0, 5):
```
Comment (romanie) [inline diplom.py]: make this min(5, len(classes_by_prob))

Comment (oksanaie) [inline diplom.py]: I don't understand what did u mean by this:
 make this min(5, len(classes_by_prob))

Comment (oksanaie) [inline diplom.py]: Here is the output, thing I don't understand is why the precision of output is higher on log. regression? Random forest gives significantly better result.

Roman-computer:Diploma oksana$ ./diplom.py --model random_forest
Model used in this run is random_forest
MAP on test data: 0.14479
Map on train data: 0.72508

Roman-computer:Diploma oksana$ ./diplom.py --model logistic_regression
Model used in this run is logistic_regression
/Library/Python/2.7/site-packages/numpy-1.11.0-py2.7-macosx-10.10-intel.egg/numpy/core/fromnumeric.py:2652: VisibleDeprecationWarning: `rank` is deprecated; use the `ndim` attribute or function instead. To find the rank of a matrix see `numpy.linalg.matrix_rank`.
  VisibleDeprecationWarning)
MAP on test data: 0.0562411111111
Map on train data: 0.0691433333333

File: diplom.py [hunk 10]
```diff
@@ -52,33 +61,31 @@ def mean_average_precision(predicted_probability_distribution,
...
 # print test_X[0], test_y[0]
 
-from sklearn import linear_model
-logistic = linear_model.LogisticRegression(C = 1, solver = 'lbfgs', 
-    multi_class = 'multinomial', max_iter = 100)
-print('LogisticRegression score: %f'
-      % logistic.fit(train_X, train_y).score(test_X, test_y))
+
+if args.model == "logistic_regression":
+    logistic = linear_model.LogisticRegression(C = 1, solver = 'lbfgs',
```
Comment (romanie) [inline diplom.py]: Please rename variable 'logistic' into 'model'

File: diplom.py [hunk 11]
```diff
@@ -52,33 +61,31 @@ def mean_average_precision(predicted_probability_distribution,
...
 
 predicted_probability_distribution_test = logistic.predict_proba(test_X)
 predicted_probability_distribution_train = logistic.predict_proba(train_X)
 
-
 MAP_test = mean_average_precision(predicted_probability_distribution_test, 
                                   logistic.classes_, 
                                   test_y)
-print MAP_test
+print "MAP on test data: %s" % MAP_test
```
Comment (romanie) [inline diplom.py]: Although, this will work because of implicit type conversion of float to string, I would suggest you to format that float by hand and limit precision by changing "%s" to "%.3f".

File: diplom.py [hunk 12]
```diff
@@ -52,33 +61,31 @@ def mean_average_precision(predicted_probability_distribution,
...
 
-from sklearn import linear_model
-logistic = linear_model.LogisticRegression(C = 1, solver = 'lbfgs', 
-    multi_class = 'multinomial', max_iter = 100)
-print('LogisticRegression score: %f'
-      % logistic.fit(train_X, train_y).score(test_X, test_y))
+
+if args.model == "logistic_regression":
+    logistic = linear_model.LogisticRegression(C = 1, solver = 'lbfgs', 
+                                               multi_class = 'multinomial', max_iter = 100)
```
Comment (romanie) [inline diplom.py]: Also, please remove spaces when passing arguments: "C = 1" => "C=1" like you did in other places so far.

File: diplom.py (hunk 11 above)
Comment (oksanaie) [inline diplom.py]: if I write this way
print "MAP on test data: %s" %.3f MAP_test
I get error: invalid syntax

Comment (oksanaie) [inline diplom.py]: or did you mean this way
print "MAP on test data: %.3f" % MAP_test
then I get the same small precision 0.056

Comment (romanie) [inline diplom.py]: Yes, that what I meant. Point is not to have higher precision - but to rather control it directly ;)

File: diplom.py (hunk 9 above)
Comment (romanie) [inline diplom.py]: Change "Map on train" to "MAP on train". Precision should be fixed now.

File: diplom.py [hunk 13]
```diff
@@ -52,33 +61,31 @@ def mean_average_precision(predicted_probability_distribution,
...
 MAP_train = mean_average_precision(predicted_probability_distribution_train, 
-                                   logistic.classes_, 
-                                   train_y) 
-print MAP_train
-# from sklearn import svm
-# clf = svm.SVC()
-# clf.fit (r, b)
-
-# print "hello"
+                                    model.classes_,
```
Comment (romanie) [inline diplom.py]: Fix indent please

File: diplom.py [hunk 14]
```diff
@@ -52,33 +61,31 @@ def mean_average_precision(predicted_probability_distribution,
 # print train_X[0], train_y[0]
```
Comment (romanie) [inline diplom.py]: Remove this outdated output.

File: diplom.py [hunk 15]
```diff
@@ -52,33 +61,31 @@ def mean_average_precision(predicted_probability_distribution,
...
-# clf = svm.SVC()
-# clf.fit (r, b)
-
-# print "hello"
+                                    model.classes_, 
+                                    train_y) 
+print "MAP on train data: %.3f" % MAP_train
 
-# # head -n1500 train.csv > minitest.txt
+# # # head -n1500 train.csv > minitest.txt
```
Comment (romanie) [inline diplom.py]: Remove these outdated comments

File: diplom.py [hunk 16]
```diff
@@ -52,33 +61,31 @@ def mean_average_precision(predicted_probability_distribution,
...
-predicted_probability_distribution_test = logistic.predict_proba(test_X)
-predicted_probability_distribution_train = logistic.predict_proba(train_X)
+if args.model == "logistic_regression":
+    model = linear_model.LogisticRegression(C=1, solver='lbfgs', 
+                                               multi_class='multinomial', max_iter=100)
+else:
+    model = RandomForestClassifier(n_estimators=100)
 
+model.fit(train_X, train_y)
+
```
Comment (romanie) [inline diplom.py:67]: delete at least one empty line here.

File: diplom.py [hunk 17]
```diff
@@ -1,5 +1,13 @@
 #!/usr/bin/python
-print "hello"
+import ml_metrics as metrics
+import argparse
+from sklearn import linear_model
+from sklearn.ensemble import RandomForestClassifier
+
+parser = argparse.ArgumentParser()
```
Comment (romanie) [inline diplom.py]: Final thing - lets move this parsing right after function definitions so the execution flow would not be interrupted by function definitions.

**Comment by romanie** [comment]
Good work!
